src/scales/scale.time.js

```diff
@@ -334,27 +334,38 @@ function generate(min, max, minor, major, capacity, options) {
  * Returns the right and left offsets from edges in the form of {left, right}.
  * Offsets are added when the `offset` option is true.
  */
-function computeOffsets(table, ticks, min, max, options) {
+function computeOffsets(table, ts, min, max, options) {
 	var left = 0;
 	var right = 0;
-	var upper, lower;
-
-	if (options.offset && ticks.length) {
-		if (!options.time.min) {
-			upper = ticks.length > 1 ? ticks[1] : max;
-			lower = ticks[0];
-			left = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
-		}
-		if (!options.time.max) {
-			upper = ticks[ticks.length - 1];
-			lower = ticks.length > 1 ? ticks[ticks.length - 2] : min;
-			right = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+	var timestamps = [];
+	var i, ilen, timestamp, length, upper, lower, divisor;
+
+	if (options.offset) {
+		[ts.labels, ts.data].forEach(function(data) {
+			if (!timestamps.length) {
+				// Remove labels outside the min/max range
+				for (i = 0, ilen = data.length; i < ilen; ++i) {
+					timestamp = data[i];
+					if (timestamp >= min && timestamp <= max) {
+						timestamps.push(timestamp);
+					}
+				}
+			}
+		});
+
+		length = timestamps.length;
+		if (length) {
+			divisor = length > 1 ? 2 : 1;
+			if (!options.time.min) {
+				upper = length > 1 ? interpolate(table, 'time', timestamps[1], 'pos') : 1;
```

Comment:
Yes, it is deduped and sorted at https://github.com/chartjs/Chart.js/blob/b9afeaf973999516cd3fdfdf4ec7c86b223d9263/src/scales/scale.time.js#L515 and https://github.com/chartjs/Chart.js/blob/b9afeaf973999516cd3fdfdf4ec7c86b223d9263/src/scales/scale.time.js#L521